---
title: "SNP Filtering"
author: "Steve Pederson"
date: "21 September 2017"
output: 
  html_document: 
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, warning = FALSE)
```

```{r}
library(tidyverse)
library(magrittr)
library(rtracklayer)
library(GenomicRanges)
library(parallel)
library(reshape2)
nCores <- min(12, detectCores() - 1)
```


# Introduction

This file takes the raw output from `Stacks` and filters the complete set of SNPs to arrive at the final list of SNPs used for downstream analysis.
The two populations are referred to below as *Gum Creek* (1996) and *Oraparinna* (2012).

## Outline of selection process for SNPs to be analysed

The filtering steps involved are:

1. Removal of SNPs located on sex chromosomes
2. Removal of *fixed* alleles, defined as `P > 0.95` in both populations and `P = 0.5` in both populations
3. Removal of SNPs within the *PstI* restriction site as these would be the result of sequencing errors
4. Removal of SNPs not observed in 1996 as these would be likely due to migration not selection
5. Removal of SNPs within the same GBS tag, with identical allele frequencies
6. Removal of SNPs called by `Stacks` to be in separate tags, but which share identical genomic co-ordinates

Beyond these steps additional criteria were applied to select SNPs of interest.
SNPs selected for downstream analysis must be:

1. Within 40kb of a gene
2. Identified in >75% of individuals within both populations

`Stacks` identifies `P` (major) and `Q` (minor) alleles within each population, and an additional proessing step was included to ensure the `P` allele in both populations was the SNP determined to be the `P` allele in the 1996 population.

# Selection of SNPs for analysis

## Genome Setup

Genome information was loaded with chromosome & scaffold identifiers from NCBI, GenBank, RefSeq and Ensembl along with the length of each chromosome/scaffold.

```{r}
chrNames <- file.path("..", "data", "chrMapping.txt") %>%
  read_delim(delim = "\t")
```

A `Seqinfo` object was formed for all downstream `GRanges` objects using Ensembl chromosome identifiers for `seqnames`.

```{r}
sq <- Seqinfo(seqnames = chrNames$Ensembl,
              seqlengths = chrNames$Length,
              isCircular = grepl("^MT$", chrNames$Ensembl), 
              genome = "OryCun2.0")
```

## Loading Genes

The get of defined genes were loaded from Ensembl Build 84

```{r}
allGenes <- file.path("..", "data","Oryctolagus_cuniculus.OryCun2.0.84.gff3.gz") %>%
  import.gff3(feature.type = "gene") 
seqinfo(allGenes) <- sq[seqnames(seqinfo(allGenes))]
```


## Loading SNP Data

```{r, sumData, cache=TRUE, dependson='chrNames'}
sumData <- file.path("..", "data", "batch_3.sumstats.tsv.gz") %>%
  gzfile() %>%
  read_delim(delim = "\t", col_names = TRUE, skip = 2) %>%
  dplyr::select(-contains("Batch"), -contains("Pi"), -contains("Fis")) %>%
  mutate(RefSeq = gsub("^gi\\|.+\\|ref\\|(.+)\\|$", "\\1", Chr),
         snpID = paste(`Locus ID`, Col, sep="_"),
         Private = as.logical(Private)) %>%
  left_join(chrNames, by = "RefSeq") %>%
  dplyr::select(Ensembl, Chromosome, BP,
                `Locus ID`, Col, snpID,
                `Pop ID`, contains("Nuc"), N, P,
                contains("Obs"), contains("Exp"),
                Private) %>%
  arrange(Chromosome, BP, snpID, `Pop ID`)
```

The `stacks` summary output was loaded, renaming chromosomes to the `chr1, chr2, ...` format.
This object contained information for `r length(unique(sumData$snpID))` SNPs across `r length(unique(sumData[["Locus ID"]]))` GBS tags.

### Conversion of SNPs to GRanges

```{r}
snp2GR <- sumData %>%
  distinct(snpID, .keep_all = TRUE) %>%
  makeGRangesFromDataFrame(seqinfo = sq,
                           keep.extra.columns=TRUE,
                           seqnames.field = "Ensembl",
                           start.field = "BP",
                           end.field = "BP",
                           ignore.strand = TRUE) %>%
  extract(, c("Locus ID", "snpID")) %>%
  set_names(.$snpID) %>%
  subset(seqnames != "X") 
```

The locations for all SNPs were converted to a `GRanges` object removing any on Chromosome `X`.
This gave an initial list of `r length(snp2GR)` SNPs across `r length(unique(snp2GR$`Locus ID`))` GBS tags.

### Selection of SNPs Near Genes

```{r}
snpNearGene <- snp2GR %>%
  resize(80001, fix = "center") %>%
  trim() %>%
  subsetByOverlaps(allGenes) %>%
  names()
```

```{r}
sumData %<>% filter(snpID %in% snpNearGene)
```

Before filtering based on population frequencies or sample size, the list of candidate SNPs was first restricted to the `r length(snpNearGene)` SNPs within $\pm 40$kb of a gene.
This represented `r length(unique(sumData[["Locus ID"]]))` GBS tags.

### FIltering Based on Data Observations

#### Identifying Fixed SNPs

```{r}
fixedSNPs <- sumData %>%
  group_by(snpID) %>%
  summarise(minP = min(P),
            maxP = max(P),
            both50 = all(minP == 0.5, maxP == 0.5)) %>%
  filter(minP > 0.95 | both50)
```

This marked `r nrow(fixedSNPs)` SNPs for removal.

#### Identifying SNPs in the Restriction Site

```{r}
rsSNPs <- filter(sumData, Col < 3) %>%
  distinct(snpID)
```

This marked a further `r nrow(rsSNPs)` SNPs for removal.


#### Identifying SNPs Not Detected in 1996

```{r}
uniq2Ora <- filter(sumData, `Pop ID` == 1, P == 1) 
```

This marked `r nrow(uniq2Ora)` SNPs for removal.

### Identifying SNPs With Low Detection Rates

```{r}
minSamp <- sumData %>% 
  group_by(`Pop ID`) %>% 
  summarise(N = max(N)) %>%
  mutate(minN = ceiling(0.75*N)) %>%
  dplyr::select(`Pop ID`, minN)
```


```{r}
lowSamp <- sumData %>%
  left_join(minSamp) %>%
  mutate(keep = N > minN) %>%
  group_by(snpID) %>%
  summarise(keep = sum(keep) == 2) %>%
  filter(!keep) %>%
  dplyr::select(snpID)
```


```{r}
rmSNPs <- unique(c(fixedSNPs$snpID,
                   rsSNPs$snpID,
                   uniq2Ora$snpID,
                   lowSamp$snpID))
```

The combination of the above three steps marked `r length(rmSnps)` for removal.

#### Identifying Non-Redundant SNPs

```{r, rmRedundant}
bestInTag <- sumData %>%
  dplyr::filter(!snpID %in% rmSNPs) %>%
  split(f = .$`Locus ID`) %>%
  mclapply(function(x){
    dcast(x[c("snpID", "Pop ID", "N")], snpID ~ `Pop ID`) %>%
      distinct(`1`, `2`, .keep_all = TRUE)
  }, mc.cores = nCores) %>%
  bind_rows %>%
  dplyr::select(snpID)
```

Where multiple SNPs were present in a tag, with identical allele frequencies and identification rates, one was selected at random.
This yielded a list of `r nrow(bestInTag)` SNPs.


#### Removal of SNPs Duplicated in Multiple Tags

```{r}
bestNoDups <- sumData %>%
  filter(snpID %in% bestInTag$snpID) %>%
  group_by(Chromosome, BP, snpID) %>%
  summarise(N = sum(N)) %>%
  ungroup() %>%
  arrange(Chromosome, BP, N) %>%
  distinct(Chromosome, BP, .keep_all = TRUE)
```

Some genomic positions were also identified in more than one GBS tag, and the ones with the highest number of identified samples were selected.
If both were identical, the one was selected at random.

```{r}
sumData %<>% filter(snpID %in% bestNoDups$snpID)
```


## Preparation of Initial Filtered List

This gave a total of `r length(unique(sumData$snpID))` SNPs across `r length(unique(sumData[["Locus ID"]]))` GBS tags as the best candidates for downstream analysis.

## Correction of P & Q Nucleotides