---
title: "SNP Analysis"
author: "Steve Pederson"
date: "24th September 2017"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 8
    fig_width: 10
    keep_md: yes
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, warning = FALSE)
```

```{r loadPackages}
library(tidyverse)
library(parallel)
library(pander)
library(scales)
library(reshape2)
library(readxl)
library(magrittr)
library(VennDiagram)
library(qqman)
library(sp)
library(ggmap)
```


```{r setParams}
nCores <- min(12, detectCores() - 1)
logit <- binomial()$linkfun
alpha <- 0.05
theme_set(theme_bw())
```


# Introduction

This analysis takes the filtered SNPs from the [previous section](02_snpFiltering) and analyses their population frequencies amongst the two populations.
The two populations are referred to below as Population `1` (*Gum Creek*, 1996) and Population `2` (*Oraparinna*, 2012).

## Outline of analysis

The analysis uses two simple models fully described by Lewis in [Genetic association studies: Design, analysis and interpretation](https://doi.org/10.1093/bib/3.2.146) to detect statistically significant differences between the two populations

1. A *full genotype* model based on **genotype counts**, which will detect changes in the structure of heterozygous or homozygous nature of genotypes
2. A *multiplicative model* based on **allele counts** in which the impact of a single copy of an allele will be tested.

- Prior to analysis, PCA was performed as a data exploration procedure
- Models were tested using Fisher's Exact Test to allow for low counts in contingency tables
- Multiple testing adjustments were performed using both *Bonferroni's method* and the *Benjamini-Hochberg method* to assess results in the context of both FWER and FDR control.

# Data Setup

```{r allData}
allData <- file.path("..", "data", "filteredSNPs.tsv.gz") %>%
  gzfile() %>%
  read_delim(delim = "\t") 
```

```{r popSizes}
popSizes <- allData %>%
  group_by(`Pop ID`) %>%
  summarise(N = max(N)) %>%
  mutate(Population = c("1996", "2012")[`Pop ID`],
         Description = c("Gum Creek", "Oraparinna")[`Pop ID`]) %>%
  ungroup() %>%
  dplyr::select(contains("Pop"), Description, N)
```

Data from the `r comma(length(unique(allData$snpID)))` SNPs selected for testing was loaded, and sample sizes formed as a simple table.
Using the observed major allele frequencies (`P`) and heterozygote frequencies (`Obs Het`), the allele frequency and genotype frequency tables were constructed.

```{r allCounts}
allCounts <- allData %>%
  mutate(A = round(P*2*N, 0), 
         B = round((1-P)*2*N, 0), 
         AB = round(`Obs Het`*N, 0), 
         AA = round((A - AB)/2, 0), 
         BB = round((B - AB)/2, 0)) %>%
  dplyr::select(snpID, `Pop ID`, A, B, AA, AB, BB)
```

The complete set of minor allele counts derived from the original genepop file was also loaded

```{r minorAlleleCounts}
minorAlleleCounts <- file.path("..", "data", "minorAlleleCounts.tsv.gz") %>%
  gzfile() %>%
  read_tsv() %>%
  as.data.frame() %>%
  column_to_rownames("sampleID") %>%
  as.matrix()
```

Populations are defined by the sample names in this object were also defined.

```{r allelePops}
allelePops <- gsub("(gc|ora|TF|Y|pt|tf).+", "\\1", rownames(minorAlleleCounts))
allelePops[!allelePops %in% c("gc", "ora")] <- "tf"
allelePops <- as.factor(allelePops)
names(allelePops) <- rownames(minorAlleleCounts)
```

In addition, the metadata from 2012 sample collection was loaded.

```{r sampleMetadata}
sampleMetadata <- file.path("..", "data", "Samples_Nov2012_GumCreek.xlsx") %>%
  read_excel(skip = 1) %>%
  mutate(ID = gsub("[Oo][Rr][Aa] ", "ora", ID))
```

GPS Locations were added to the metadata.

```{r}
gpsPoints <-file.path("..", "data", "GPS_Locations.xlsx") %>%
  read_excel() %>%
  dplyr::select(ID = Sample, ends_with("tude")) %>%
  mutate(ID = gsub("[Oo][Rr][Aa] ([0-9ABC]*)", "ora\\1", ID))
sampleMetadata %<>% left_join(gpsPoints, by = "ID")
```



# Principal Component Analysis

## Missing Values

In order to perform PCA, missing values must be either imputed or the entire sample/SNPs must be removed.

```{r missingBySample}
missingBySample <- minorAlleleCounts %>%
  apply(MARGIN = 1, function(x){mean(is.na(x))}) 
```

`r sum(missingBySample > 1/3)` samples were missing information for more than 1/3 of SNPs and for the purposes of PCA, these samples were removed from the dataset.


```{r missingByAllele}
missingByAllele <- minorAlleleCounts %>%
  extract(missingBySample <= 1/3,) %>%
  apply(MARGIN = 2, function(x){mean(is.na(x))}) 
```

Of these remaining samples `r sum(missingByAllele < 0.05)` alleles were identified in >95% of individuals, and this subset of the data was then chosen as the basis for PCA.

```{r dataForPCA}
dataForPCA <- minorAlleleCounts[missingBySample <= 1/3, missingByAllele < 0.05]
```

Missing values were then randomly sampled using the distribution of alleles in the remaining samples.
Population structure was not taken into account at this point so as to not artificially inflate correlations within populations.

```{r dataForPCA_missing}
dataForPCA %<>%
  apply(MARGIN = 2, function(x){
    missing <- is.na(x)
    if (sum(missing) > 0) {
      x[missing] <- sample(x[!missing], sum(missing),replace = TRUE)
    }
    x
  })
```

After this process it was noted that there was no allelic differences amongst the remaining samples for `r sum(apply(dataForPCA, MARGIN = 2, function(x){length(unique(x))}) == 1)` SNPs, implying that the observed variants at these loci were contained within the samples with large numbers of missing data.
These loci were also removed before performing PCA.

```{r dataForPCA_subset}
dataForPCA %<>% extract(, apply(., MARGIN = 2, function(x){length(unique(x))}) > 1)
```


## PCA Including The Outgroup Samples

```{r outgroupPCA}
set.seed(53)
outgroupPCA <- dataForPCA %>%
  prcomp(center = TRUE)
```

The two Gum Creek populations (1996 & 2012) clearly showed differences to the outgroup, however a strong "tail" was noted for some of the 2012 population along PC2.

The samples forming this tail (i.e. PC2 > 5 & PC3 < 0 ) were identified and the collection region for these samples was investigated, using the GPS collection point.
The vast majority were found to come from the central collection region, and this sub-population was added to the PCA plot.


```{r pcaForPlot}
pcaForPlot <- outgroupPCA$x %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  as_data_frame() %>%
  dplyr::select(sampleID, PC1, PC2, PC3) %>%
  mutate(Population = allelePops[sampleID]) %>%
  rowwise() %>%
  mutate(Population = if_else(Population == "gc", "Gum Creek (1996)",
                              if_else(Population == "ora", "Oraparinna (2012)", "Turretfield (2010)"))) %>%
  left_join(sampleMetadata, by = c("sampleID" = "ID")) %>%
  mutate(Population = if_else(Population == "Oraparinna (2012)",
                              if_else(PC3 < 0 && PC2 > 5, "Oraparinna Central (2012)", "Oraparinna Outer (2012)" ), Population))
```

```{r finalPCA, echo = FALSE, fig.cap = "PCA for all samples including the outgroup and indicating the sample collection region for the 2012 samples."}
p12 <- ggplot(pcaForPlot, aes(PC1, PC2, colour = Population)) +
  geom_point(size = 3) +
  scale_colour_manual(values = c(rgb(0, 0, 0, 0.7),
                                 rgb(0.8, 0.2, 0.2, 0.9),
                                 rgb(0.2, 0.8, 0.2, 0.9),
                                 rgb(0.5, 0.5, 1, 0.7))) +
  theme(text = element_text(size = 12))
p13 <- ggplot(pcaForPlot, aes(PC1, PC3, colour = Population)) +
  geom_point(size = 3) +
  scale_colour_manual(values = c(rgb(0, 0, 0, 0.7),
                                 rgb(0.8, 0.2, 0.2, 0.9),
                                 rgb(0.2, 0.8, 0.2, 0.9),
                                 rgb(0.5, 0.5, 1, 0.7))) +
  guides(colour = FALSE) +
  theme(text = element_text(size = 12))
p23 <- ggplot(pcaForPlot, aes(PC2, PC3, colour = Population)) +
  geom_point(size = 3) +
  scale_colour_manual(values = c(rgb(0, 0, 0, 0.7),
                                 rgb(0.8, 0.2, 0.2, 0.9),
                                 rgb(0.2, 0.8, 0.2, 0.9),
                                 rgb(0.5, 0.5, 1, 0.7))) +
  guides(colour = FALSE) +
  theme(text = element_text(size = 12))
vp12 <- viewport(0, 0, 0.71, 0.5, just = c(0, 0))
vp13 <- viewport(0, 0.5, 0.5, 0.5, just = c(0, 0))
vp23 <- viewport(0.5, 0.5, 0.5, 0.5, just = c(0, 0))
grid.newpage()
print(p12, vp = vp12)
print(p13, vp = vp13)
print(p23, vp = vp23)
```


```{r map1, echo=FALSE, fig.cap="Collection points for all 2012 samples with colours showing sub-populations defined by PCA analysis.", fig.height=7, fig.width=7}
loc <- c(range(sampleMetadata$Longitude) %>% mean,
         range(sampleMetadata$Latitude) %>% mean)
ggMap <- get_map(loc, zoom = 12, maptype = "hybrid", source = "google")
ggmap(ggMap) +
  geom_point(data= pcaForPlot, aes(Longitude, Latitude, colour = Population)) +
  scale_colour_manual(values = c(rgb(0, 0, 0, 0.7),
                                 rgb(0.8, 0.2, 0.2, 0.9),
                                 rgb(0.2, 0.8, 0.2, 0.9),
                                 rgb(0.5, 0.5, 1, 0.7))) +
  theme_bw() +
  guides(colour = FALSE) +
  labs(x = "Longitude",
       y = "Latitude")
```

```{r map2, echo=FALSE, fig.cap="Zoomed-in view of the central region for 2012 samples with colours showing sub-populations defined by PCA analysis.", fig.height=7, fig.width=7}
zoomLoc <- c(138.753, -31.242)
get_map(zoomLoc, zoom = 15, maptype = "hybrid", source = "google") %>%
  ggmap() +
  geom_point(data= pcaForPlot, aes(Longitude, Latitude, colour = Population),
             size = 3) +
  scale_colour_manual(values = c(rgb(0, 0, 0, 0.7),
                                 rgb(0.8, 0.2, 0.2, 0.9),
                                 rgb(0.2, 0.8, 0.2, 0.9),
                                 rgb(0.5, 0.5, 1, 0.7))) +
  theme_bw() +
  guides(colour = FALSE) +
  labs(x = "Longitude",
       y = "Latitude")
```

# Phylogenetic Analysis

## Data Export For Generation of Phylogenies

The SNP information for the original `r ncol(minorAlleleCounts)` alleles was then exported for phylogenetic analysis.
The same subset of `r nrow(pcaForPlot)` samples was used for this analysis as was used for the PCA analysis above.

```{r}
alleles <- allData %>%
  dplyr::select(snpID, contains("Nuc")) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  rowwise() %>%
  mutate(AA = paste0(`P Nuc`, `P Nuc`),
         AB = paste0(`P Nuc`, `Q Nuc`),
         BB = paste0(`Q Nuc`, `Q Nuc`)) %>%
  ungroup %>%
  dplyr::select(snpID, AA, AB, BB) %>%
  as.data.frame() %>%
  column_to_rownames("snpID") %>%
  as.matrix()
```

```{r}
gzOut <- file.path("..", "data", "snpForPhylo.tsv.gz") %>% gzfile("w")
colnames(minorAlleleCounts) %>% 
  vapply(function(x){
    alleles[x,  minorAlleleCounts[,x] + 1]
    }, character(nrow(minorAlleleCounts))) %>%
  as_data_frame() %>%
  mutate(sampleID = rownames(minorAlleleCounts)) %>%
  filter(sampleID %in% pcaForPlot$sampleID) %>%
  mutate(Population = gsub("gc.+", "Gum Creek (1996)", sampleID),
         Population = gsub("(TF|Y|pt|tf).+", "Turretfield (2010)", Population),
         Population = ifelse(sampleID %in% filter(pcaForPlot, grepl("Central", Population))$sampleID,
                             "Oraparinna Central (2012)", Population),
         Population = gsub("ora.+", "Oraparinna Outer (2012)", Population),
         colour = ifelse(grepl("Turretfield", Population), rgb(0.5, 0.5, 1, 0.7), ""),
         colour = ifelse(grepl("Gum", Population), rgb(0, 0, 0, 0.7), colour),
         colour = ifelse(grepl("Central", Population), rgb(0.8, 0.2, 0.2, 0.9), colour),
         colour = ifelse(grepl("Outer", Population), rgb(0.2, 0.8, 0.2, 0.9), colour)) %>%
  dplyr::select(sampleID, Population, colour, everything()) %>%
  write_tsv(gzOut)
close(gzOut)
```


# Analysis

## Removal of SNPs Associated with Collection Region

The structure observed within the 2012 population in the PCA could possibly be explained by recent migration into this region.
As the samples collected in the outer regions appeared very similar to the 1996 population in the above plots, this would possibly indicate this a very recent event as the genetic influence of this has not spread through the wider area.
Although this may be due to other factors such as sampling bias, this structure was addressed by identifying SNPs which showed an association with the sub-populations identified by PCA analysis.
In this way, any candidate SNPs obtained below will be less impacted by this structure, and will be more reflective of the intended variable under study, as opposed to any internal structure of the 2012 population.

```{r oraRegions}
oraRegions <- pcaForPlot %>% 
  filter(grepl("2012", Population)) %>% 
  mutate(Central = grepl("Central", Population)) %>% 
  dplyr::select(sampleID, Central) %>%
  as.data.frame() %>%
  column_to_rownames("sampleID") 
```


### Testing for Structure in 2012

This model tests:  
H<sub>0</sub>: No association between genotypes and collection region  
H<sub>A</sub>: An association exists between genotypes and collection region

```{r ora}
ora <- grep("ora", rownames(minorAlleleCounts), value = TRUE)
regionResults <- minorAlleleCounts %>%
  extract(ora,) %>%
  apply(MARGIN = 2, function(x){
    mat <- table(oraRegions[ora, "Central"], x)
    if (ncol(mat) > 1){
      fisher.test(mat)$p.value
    }
    else{
      NA
    }
  })
```

A total of `r sum(regionResults < 0.05, na.rm = TRUE)` SNPs were detected as showing a significant association between genotype and the collection region.
Under H<sub>0</sub>, the number expected using &#945; = 0.05 would be `r floor(0.05*ncol(minorAlleleCounts))`, and as this number was approximately double that expected, this was taken as evidence of this being a genuine point of concerning this data.

Type II errors were of principle concern in this instance, and as such every SNP with p < 0.05 in the above test was excluded from downstream analysis.

```{r regionSNPs}
regionSNPs <- names(which(regionResults < 0.05))
regionSNPs %>% writeLines(file.path("..", "results", "regionSNPs.txt"))
```

Under this additional filtering step, **the original set of `r length(regionResults)` SNPs will be reduced to `r length(regionResults) - length(regionSNPs)`** for testing by genotype and allele frequency.


### Verification Of Removal

In order to verify that the removal of the above SNPs removed the undesired population structure from the 2012 population, the above PCA was repeated, excluding the SNPs marked for removal.
The previous structure noted in the data was no longer evident, and as such, these SNPs were marked for removal during analysis by genotype and allele frequency. 

```{r reducedPCA}
reducedPCA <- dataForPCA %>%
  extract(, !colnames(.) %in% regionSNPs) %>%
  prcomp()
```

```{r plotReducedPCA, echo = FALSE, fig.cap="Principal Component Analysis after removal of SNPs showing evidence of structure within the 2012 population.", fig.height=5, fig.width=7}
reducedPCA$x %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  as_data_frame() %>%
  dplyr::select(sampleID, PC1, PC2, PC3) %>%
  mutate(Population = allelePops[sampleID]) %>%
  rowwise() %>%
  mutate(Population = if_else(Population == "gc", "Gum Creek (1996)",
                              if_else(Population == "ora", "Oraparinna (2012)", "Turretfield (2010)"))) %>%
  left_join(sampleMetadata, by = c("sampleID" = "ID")) %>%
  mutate(Central = sampleID %in% filter(pcaForPlot, grepl("Central", Population))$sampleID,
         Population = if_else(Population == "Oraparinna (2012)",
                              if_else(Central, "Oraparinna Central (2012)", "Oraparinna Outer (2012)" ), Population)) %>%
  ggplot(aes(PC1, PC2, colour = Population)) +
  geom_point(size = 3) +
  scale_colour_manual(values = c(rgb(0, 0, 0, 0.7),
                                 rgb(0.8, 0.2, 0.2, 0.9),
                                 rgb(0.2, 0.8, 0.2, 0.9),
                                 rgb(0.5, 0.5, 1, 0.7))) 
```


## Genotype Frequency Model

This model tests:  
H<sub>0</sub>: No association between genotypes and populations  
H<sub>A</sub>: An association exists between genotypes and populations

```{r genotypeResults}
genotypeResults <- allCounts %>%
  filter(!snpID %in% regionSNPs) %>%
  split(f = .$snpID) %>%
  mclapply(function(x){
    mat <- as.matrix(x[,c("AA", "AB",  "BB")])
    ft <- fisher.test(mat)
    data_frame(snpID = unique(x$snpID),
               p = ft$p.value)
  }, mc.cores = nCores) %>%
  bind_rows() %>%
  mutate(adjP = p.adjust(p, method = "bonferroni"),
         FDR = p.adjust(p, method = "fdr")) %>%
  arrange(p)
```


Under the full genotype model:

- `r nrow(filter(genotypeResults, adjP < alpha))` genotypes were detected as being significantly associated with the two populations when controlling the FWER at &#945; = `r alpha`
- `r nrow(filter(genotypeResults, FDR < alpha))` genotypes were detected as being significantly associated with the two populations when controlling the FDR at &#945; = `r alpha`
- If controlling the FDR at 10% however, a total of `r nrow(filter(genotypeResults, FDR < 0.1))` genotypes were considered as potentially associated with the population structure 
- For the most highly ranked SNP (`r genotypeResults$snpID[[1]]`), the minor allele has been completely lost in the 2012 population

```{r,echo=FALSE}
genotypeResults %>%
  filter(p < 1e-04) %>%
  left_join(allData) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  dplyr::select(Chr, BP, snpID, p, FDR) %>%
  pander(justify = "rrlrr",
         style = "rmarkdown",
         big.mark = ",",
         caption = "SNPs with raw p-values < 1e-04 when analysing by genotype. All SNPs were considered significant using an FDR < 0.1")
```

```{r manhattanGenotype, echo = FALSE, fig.height=5, fig.width=8, fig.cap = "Manhattan plot showing results for all SNPs on chromosomes 1:21 for analysis by genotype. The horizontal line indicates the cutoff for an FDR of 10%"}
hLine <- max(filter(genotypeResults, FDR < 0.1)$p)
genotypeResults %>%
  left_join(allData) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  mutate(Chr = as.numeric(Chr)) %>%
  filter(!is.na(Chr)) %>%
  dplyr::select(SNP = snpID, CHR = Chr, BP, P = p) %>%
  manhattan(suggestiveline = FALSE, 
            genomewideline = -log10(hLine))
```


## Allele Frequency Model

This model tests:  
H<sub>0</sub>: No association between allele frequencies and populations  
H<sub>A</sub>: An association exists between allele frequencies  and populations

```{r alleleResults}
alleleResults <- allCounts %>%
  filter(!snpID %in% regionSNPs) %>%
  split(f = .$snpID) %>%
  mclapply(function(x){
    mat <- as.matrix(x[,c("A", "B")])
    ft <- fisher.test(mat)
    data_frame(snpID = unique(x$snpID),
               p = ft$p.value)
  }, mc.cores = nCores) %>%
  bind_rows() %>%
  mutate(adjP = p.adjust(p, method = "bonferroni"),
         FDR = p.adjust(p, method = "fdr")) %>%
  arrange(p)
```

Under this model:

- `r nrow(filter(alleleResults, adjP < alpha))` SNP alleles were detected as being significantly associated with the two populations when controlling the FWER at &#945; = `r alpha`.
However, as these SNPs were within 21nt of each other, this may represent the same haplotype 
- `r nrow(filter(alleleResults, FDR < alpha))` SNP alleles were detected as being significantly associated with the two populations when controlling the FDR at &#945; = `r alpha`
- extending the FDR to 10% yielded `r nrow(filter(alleleResults, FDR < 0.1))` SNP alleles


```{r,echo=FALSE}
alleleResults %>%
  filter(FDR < alpha) %>%
  left_join(allData) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  dplyr::select(Chr, BP, snpID, p, adjP, FDR) %>%
  pander(justify = "rrlrrr",
         big.mark = ",",
         style = "rmarkdown",
         caption = paste("SNPs considered as significant when analysing by genotype using an FDR cutoff of", alpha))
```


```{r manhattanAllele, echo = FALSE, fig.height=5, fig.width=8, fig.cap = "Manhattan plot showing results for all SNPs on chromosomes 1:21 when analysing by allele frequencies. The horizontal line indicates the cutoff for an FDR of 10%, with SNPs considered significant under the Bonferroni adjustnt shown in green."}
hLine <- max(filter(alleleResults, FDR < 0.1)$p)
hLight <- filter(alleleResults, adjP< 0.05)$snpID
alleleResults %>%
  left_join(allData) %>%
  distinct(snpID, .keep_all = TRUE) %>%
  mutate(Chr = as.numeric(Chr)) %>%
  filter(!is.na(Chr)) %>%
  dplyr::select(SNP = snpID, CHR = Chr, BP, P = p) %>%
  manhattan(suggestiveline = FALSE, 
            genomewideline = -log10(hLine),
            highlight = hLight)
```

## FLK Analysis

The FLK analysis performed separately was also included in order to compare the differing approaches.
It should be noted that this analytic approach is very closely related to analysis by allele frequency, however, instead of Fisher's Exact Test a Chi-squared model is utilised incorporating genetic distances to ccount for genetics drift.

```{r}
flkResults <- file.path("..", "results", "flkResults.tsv") %>%
  read_tsv()
```

## Bayescan Analysis

```{r snpsForBayescan}
snpsForBayescan <- file.path("..", "data", "filteredSNPs.genepop.gz") %>%
  gzfile() %>%
  readLines(n = 2) %>%
  extract2(2) %>%
  strsplit(",") %>%
  extract2(1)
```

The analysis tool Bayescan was also run on the set of `r comma(length(snpsForBayescan))` SNPs exported as a `genepop` file after the previous filtering steps.
The program was run using an FDR of 0.05 with an additional run setting the FDR to 0.1.
Any SNPs detected above as tracking with the internal structure of the 2012 population were discarded from the results.

```{r bayes05}
bayes05 <- file.path("..", "results", "BayOut_FDR5.csv") %>%
  read_csv() %>%
  mutate(snpID = snpsForBayescan[SNP]) %>%
  dplyr::select(snpID, everything(), -SNP) %>%
  filter(!snpID %in% regionSNPs)
```


```{r bayes10}
bayes10 <- file.path("..", "results", "BayOut_FDR10.csv") %>%
  read_csv() %>%
  mutate(snpID = snpsForBayescan[SNP])%>%
  dplyr::select(snpID, everything(), -SNP) %>%
  filter(!snpID %in% regionSNPs)
```

## Comparison of Results

```{r sigSNPs}
fdr <- c(genotype = 0.1, allele = 0.05, flk = 0.05)
sigSNPs <- c(filter(genotypeResults,FDR < fdr["genotype"])$snpID,
             filter(alleleResults, FDR < fdr["allele"])$snpID,
             filter(flkResults, FDR < fdr["flk"])$snpID,
             bayes10$snpID) %>%
  unique %>%
  as.data.frame() %>%
  set_names("snpID") %>%
  mutate(genotype = snpID %in% filter(genotypeResults,FDR < fdr["genotype"])$snpID,
         allele = snpID %in% filter(alleleResults,FDR < fdr["allele"])$snpID,
         flk = snpID %in% filter(flkResults, FDR < fdr["flk"])$snpID,
         bayes = snpID %in% bayes10$snpID)
```

- It was noted when comparing results from FLK and analysis by allele frequency alone that all SNPs from the allele frequency analysis were detected by either FLK or the genotpe analysis.
As such this analysis was disregarded going forward.
- Similarly, no novel SNPs were detected by Bayescan, and all in the list down to an FDR of 0.1 were included in either the analysis using FLK or the full genotypes.
As such, the Bayescan analysis was also disregarded going forward.


```{r vennSNPs, echo = FALSE, fig.cap="Overlap between the lists of SNPs considered as associated with the different populations under either of the two analytic approaches, using an FDR of 0.1 for 'Analysis by Genotype' and Bayescan, with an FDR of 0.5 for 'Analysis by Allele' and FLK.", fig.width=6, fig.height=5, results='hide', fig.show='asis'}
grid.newpage()
draw.quad.venn(
  area1 = sum(sigSNPs$genotype),
  area2 = sum(sigSNPs$allele),
  area3 = sum(sigSNPs$flk),
  area4 = sum(sigSNPs$bayes),
  n12 = filter(sigSNPs, genotype, allele) %>% nrow(),
  n13 = filter(sigSNPs, genotype, flk) %>% nrow(),
  n14 = filter(sigSNPs, genotype, bayes) %>% nrow(),
  n23 = filter(sigSNPs, allele, flk) %>% nrow(),
  n24 = filter(sigSNPs, allele, bayes) %>% nrow(),
  n34 = filter(sigSNPs, flk, bayes) %>% nrow(),
  n123 = filter(sigSNPs, genotype, allele, flk) %>% nrow(),
  n124 = filter(sigSNPs, genotype, allele, bayes) %>% nrow(),
  n134 = filter(sigSNPs, genotype, flk, bayes) %>% nrow(),
  n234 = filter(sigSNPs, allele, flk, bayes) %>% nrow(),
  n1234 = filter(sigSNPs, genotype, allele, flk, bayes) %>% nrow(),
  category = c("Analysis by\nGenotype", "Analysis by\nAllele","FLK", "Bayescan"), 
  scaled = FALSE
  )
```

### SNPs Associated With Populations Under Both Approaches

The list of SNPs detected as associated with the population structure under both FLK and analysis by genotype is given below.

```{r, echo = FALSE}
sigSNPs %>%
  filter(genotype, flk) %>%
  dplyr::select(snpID) %>%
  left_join(allData) %>%
  mutate(SNP = paste(`P Nuc`, `Q Nuc`, sep = "/"),
         Population = popSizes$Population[`Pop ID`]) %>%
  arrange(Chr, BP, Population) %>%
  dplyr::select(Chr, BP, snpID, SNP, Population, P) %>%
  split(f = .$snpID) %>%
  lapply(function(x){
    dcast(x, Chr + BP + snpID ~ Population, value.var = "P") %>%
      mutate(`Change in log(OR)` = diff(c(logit(`1996`), logit(`2012`))))
  }) %>%
  bind_rows() %>%
  arrange(Chr, BP) %>%
  left_join(genotypeResults) %>%
  dplyr::select(Chr, BP, snpID, `Change in log(OR)`, P_1996 = `1996`, P_2012 = `2012`, Genotype_p = p) %>%
  left_join(flkResults) %>%
  dplyr::select(snpID, Chr, BP, `Change in log(OR)`, P_1996, P_2012, Genotype_p, FLK_p = F.LK.p.val) %>%
  pander(justify = "llrrrrrr",
         style = "rmarkdown",
         big.mark = ",",
         split.table = Inf,
         caption = paste("Summary of changes in the major (P) allele between the two timepoints.",
                         "Changes in the log Odds ratio of observing the major allele are given,",
                         "along with estimated population frequencies.",
                         "Results from testing by genotype or FLK are given as raw p-values.",
                         "All SNPs were considered as differentially associated with the two populations under both analyses to an FDR of 10\\% (genotype) or 5\\% (FLK)"))
```


```{r genotypesBothModels, echo = FALSE,fig.cap = "Estimated genotype frequencies for SNPs found to be significant under both models. In each case the `A` allele represents the major allele in the 1996 population.", fig.width=9, fig.height=7}
sigSNPs %>%
  filter(genotype, flk) %>%
  dplyr::select(snpID) %>%
  left_join(allData) %>%
  left_join(allCounts) %>%
  dplyr::select(snpID, Chr, BP, `Pop ID`, AA, AB, BB) %>%
  left_join(popSizes) %>%
  melt(id.vars = c("snpID", "Chr", "BP", "Pop ID", "Population", "Description", "N"),
       variable.name = "Genotype",
       value.name  ="Count") %>%
  group_by(snpID, Chr, BP, `Pop ID`, Description) %>%
  mutate(Frequency = Count / sum(Count)) %>%
  ggplot(aes(x = Genotype, y = Frequency, fill = Population, group = Population)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_manual(values = c(rgb(0.1, 0.1, 0.1), rgb(0.2, 0.2, 0.8))) +
  facet_wrap(~snpID) 
```


### SNPs Associated With Populations Under Analysis Using FLK Only


```{r, echo = FALSE}
sigSNPs %>%
  filter(!genotype, flk) %>%
  dplyr::select(snpID) %>%
  left_join(allData) %>%
  mutate(SNP = paste(`P Nuc`, `Q Nuc`, sep = "/"),
         Population = popSizes$Population[`Pop ID`]) %>%
  arrange(Chr, BP, Population) %>%
  dplyr::select(Chr, BP, snpID, SNP, Population, P) %>%
  split(f = .$snpID) %>%
  lapply(function(x){
    dcast(x, Chr + BP + snpID ~ Population, value.var = "P") %>%
      mutate(`Change in log(OR)` = diff(c(logit(`1996`), logit(`2012`))))
  }) %>%
  bind_rows() %>%
  arrange(Chr, BP) %>%
  left_join(flkResults) %>%
  dplyr::select(Chr, BP, snpID, `Change in log(OR)`, P_1996 = `1996`, P_2012 = `2012`, FLK_p = F.LK.p.val) %>%
  pander(justify = "llrrrrr",
         style = "rmarkdown",
         big.mark = ",",         
         split.table = Inf,
         caption = paste("Summary of changes in the major (P) allele between the two timepoints.",
                         "Changes in the log Odds ratio of observing the major allele are given,",
                         "along with estimated population frequencies.",
                         "Results from testing by allele count are given as raw p-values.",
                         "All SNPs were considered as differentially associated with the two populations under FLK analysis to an FDR of 5\\%."))
```

```{r flkModelOnly, echo = FALSE,fig.cap = "Estimated allele frequencies for SNPs found to be significant under analysis by allele only. In each case the `A` allele represents the major allele in the 1996 population.", fig.width=9, fig.height=9}
sigSNPs %>%
  filter(!genotype, flk) %>%
  dplyr::select(snpID) %>%
  left_join(allData) %>%
  left_join(allCounts) %>%
  dplyr::select(snpID, Chr, BP, `Pop ID`, A, B) %>%
  mutate(A = A / (A + B),
         B = 1 - A) %>%
  left_join(popSizes) %>%
  melt(id.vars = c("snpID", "Chr", "BP", "Pop ID", "Population", "Description", "N"),
       variable.name = "Allele",
       value.name  ="Frequency") %>%
  ggplot(aes(x = Allele, y = Frequency, fill = Population)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_manual(values = c(rgb(0.1, 0.1, 0.1), rgb(0.2, 0.2, 0.8))) +
  facet_wrap(~snpID) 
```



### SNPs Associated With Populations Under Analysis By Genotype Only


```{r, echo = FALSE}
sigSNPs %>%
  filter(genotype, !flk) %>%
  dplyr::select(snpID) %>%
  left_join(allData) %>%
  mutate(SNP = paste(`P Nuc`, `Q Nuc`, sep = "/"),
         Population = popSizes$Population[`Pop ID`]) %>%
  arrange(Chr, BP, Population) %>%
  dplyr::select(Chr, BP, snpID, SNP, Population, `Obs Het`) %>%
  split(f = .$snpID) %>%
  lapply(function(x){
    dcast(x, Chr + BP + snpID ~ Population, value.var = "Obs Het") %>%
      mutate(`Change in log(OR)` = diff(c(logit(`1996`), logit(`2012`))))
  }) %>%
  bind_rows() %>%
  arrange(Chr, BP) %>%
  left_join(genotypeResults) %>%
  dplyr::select(Chr, BP, snpID, `Change in log(OR)`, P_1996 = `1996`, P_2012 = `2012`, Genotype_p = p) %>%
  pander(justify = "llrrrrr",
         style = "rmarkdown",
         big.mark = ",",         
         split.table = Inf,
         caption = paste("Summary of changes in heterozygosity between the two timepoints.",
                         "Changes in the log Odds ratio of observing heterozygotes are given,",
                         "along with estimated population-level heterozygote frequencies.",
                         "Results from testing by genotype count are given as raw p-values.",
                         "All SNPs were considered as differentially associated with the two populations under the genotype count analysis to an FDR of 10\\%,",
                         "However, no significant changes in allele frequencies were detected using an FDR of 5\\% for FLK analysis."))
```

```{r genotypeModelOnly, echo = FALSE,fig.cap = "Estimated genotype frequencies for SNPs found to be significant under analysis by genotype only. In each case the `A` allele represents the major allele in the 1996 population.", fig.width=9, fig.height=7}
sigSNPs %>%
  filter(genotype, !flk) %>%
  dplyr::select(snpID) %>%
  left_join(allData) %>%
  left_join(allCounts) %>%
  dplyr::select(snpID, Chr, BP, `Pop ID`, AA, AB, BB) %>%
  left_join(popSizes) %>%
  melt(id.vars = c("snpID", "Chr", "BP", "Pop ID", "Population", "Description", "N"),
       variable.name = "Genotype",
       value.name  ="Count") %>%
  group_by(snpID, Chr, BP, `Pop ID`, Description) %>%
  mutate(Frequency = Count / sum(Count)) %>%
  ggplot(aes(x = Genotype, y = Frequency, fill = Population)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_manual(values = c(rgb(0.1, 0.1, 0.1), rgb(0.2, 0.2, 0.8))) +
  facet_wrap(~snpID) 
```


# Output

Both sets of results were output as `genotypeResults.tsv` and `alleleResults.tsv`

```{r}
alleleResults %>% 
  left_join(allData) %>% 
  distinct(snpID, .keep_all = TRUE) %>% 
  dplyr::select(snpID, Chr, BP, p, adjP, FDR) %>%
  write_tsv(file.path("..", "results", "alleleResults.tsv"))
```
```{r}
genotypeResults %>% 
  left_join(allData) %>% 
  distinct(snpID, .keep_all = TRUE) %>% 
  dplyr::select(snpID, Chr, BP, p, adjP, FDR) %>%
  write_tsv( file.path("..", "results", "genotypeResults.tsv"))
```


```{r, echo=FALSE, results='asis'}
pander(sessionInfo()) 
```

