---
title: "Analysis Of Genes Associated With Significant SNPs"
author: "Steve Pederson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 8
    fig_width: 10
    keep_md: yes
    toc: yes
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE, warning = FALSE)
```

```{r loadPackages}
library(tidyverse)
library(rtracklayer)
library(GenomicRanges)
library(biomaRt)
library(GO.db)
library(parallel)
library(magrittr)
library(pander)
library(scales)
library(reshape2)
library(stringr)
```


```{r setParams}
nCores <- min(12, detectCores() - 1)
theme_set(theme_bw())
evalLongChunks <- FALSE
```

# Introduction

This analysis takes the SNPs found to be associated with the two populations and determines which genes are nearby.
An analysis of the GO terms associated with these SNPS is also undertaken.

## Outline of analysis

This file covers three main areas

1. Preparation of the complete mappings from Ensembl gene IDs to GO terms
2. Finding SNPs within and near genes
3. Identification of enriched GO terms using Fisher's Exact Test on the sets of genes within 2kb and 40kb of the significant SNPs

# Data Setup

## Genes

`GenomicRanges` were formed defining all genes and exons according to the gene models in Build 84 of Ensembl.

```{r gffInfo}
gffFile <- file.path("..", "data", "Oryctolagus_cuniculus.OryCun2.0.84.gff3.gz") 
ensGenes <- import.gff(gffFile, feature.type = "gene", sequenceRegionsAsSeqinfo = TRUE) %>%
  sortSeqlevels() 
ensExons <-  import.gff(gffFile, feature.type = "exon", sequenceRegionsAsSeqinfo = TRUE) %>%
  sortSeqlevels()
```

## GO Terms

### Mapping Genes to GO Terms

- The set of terms defining the root nodes was initially formed

```{r rootGO}
rootGO <- list(
  BP = "GO:0008150",
  CC = "GO:0005575",
  MF = "GO:0003674"
  )
```

- The complete set of GO IDs mapping to these genes was then obtained from biomaRt.

```{r biomaRt, eval = evalLongChunks}
mart <- useEnsembl("ensembl", dataset ="ocuniculus_gene_ensembl", version = "84")
ens2Go <- getBM(attributes = c("ensembl_gene_id","go_id", "namespace_1003"), 
                filters = "ensembl_gene_id", 
                values = ensGenes$gene_id, 
                mart = mart) %>%
  filter(go_id != "") %>% 
  dplyr::rename(ontology = namespace_1003) %>%
  mutate(ontology = c(biological_process = "BP", 
                      cellular_component = "CC", 
                      molecular_function = "MF")[ontology])
```

- GO Terms in the biomaRt results which failed to map to the database were removed.

```{r ens2Go, eval = evalLongChunks}
ens2Go %<>% split(f = .$ontology)
ens2Go$BP %<>% filter(go_id %in% get(rootGO$BP, GOBPOFFSPRING))
ens2Go$CC %<>% filter(go_id %in% get(rootGO$CC, GOCCOFFSPRING))
ens2Go$MF %<>% filter(go_id %in% get(rootGO$MF, GOMFOFFSPRING))
ens2Go %<>% bind_rows()
```
- The file was then saved to disk and became the reference database for subsequent analyses.

```{r writeEns2Go, eval = evalLongChunks}
gzOut <- file.path("..", "data", "ens2GO.biomaRt.84.tsv.gz") %>% gzfile("w")
write_delim(ens2Go, gzOut, delim ="\t")
close(gzOut)
```

```{r}
ens2Go <- file.path("..", "data","ens2GO.biomaRt.84.tsv.gz") %>%
  gzfile() %>%
  read_tsv()
```

### Building the Complete Set of Mappings

- Paths were formed mapping all GO Terms obtained from `biomaRt` back to the root nodes.

```{r allGOAncestor, eval = evalLongChunks}
allGOAncestor <- list(
  BP = filter(ens2Go, ontology == "BP") %>%
    extract2("go_id") %>%
    unique() %>%
    sapply(get, GOBPANCESTOR, simplify = FALSE),
  CC = filter(ens2Go, ontology == "CC") %>%
    extract2("go_id") %>%
    unique() %>%
    sapply(get, GOCCANCESTOR, simplify = FALSE),
  MF = filter(ens2Go, ontology == "MF") %>%
    extract2("go_id") %>%
    unique() %>%
    sapply(get, GOMFANCESTOR, simplify = FALSE)
)
```

- Each gene was then mapped to the complete set of GO terms and saved in an external table.

```{r ALLGO2ENS, eval = evalLongChunks}
ALLGO2ENS <- ens2Go %>%
  split(f = .$ensembl_gene_id) %>%
  mclapply(function(x){
    g <- unique(x$ensembl_gene_id)
    BP <- allGOAncestor$BP[filter(x, ontology == "BP")$go_id] %>%
      unlist() %>%
      c(filter(x, ontology == "BP")$go_id) %>%      
      unique()
    CC <- allGOAncestor$CC[filter(x, ontology == "CC")$go_id] %>%
      unlist() %>%
      c(filter(x, ontology == "CC")$go_id) %>%      
      unique()
    MF <- allGOAncestor$MF[filter(x, ontology == "MF")$go_id] %>%
      unlist() %>%
      c(filter(x, ontology == "MF")$go_id) %>%
      unique() 
    data_frame(ensembl_gene_id = g, go_id = c(BP, CC, MF))
  },mc.cores = nCores) %>%
  bind_rows() %>%
  filter(go_id != "all") %>%
  mutate(ontology = Ontology(go_id))
```

```{r writeALLGO2ENS, eval = evalLongChunks}
gzOut <- file.path("..", "data", "ALLGO2ENS.tsv.gz") %>% gzfile("w")
write_delim(ALLGO2ENS, gzOut, delim = "\t")
close(gzOut)
```

```{r reloadALLGO2ENS}
ALLGO2ENS <- file.path("..", "data", "ALLGO2ENS.tsv.gz") %>%
  gzfile() %>%
  read_tsv()
singleGeneGO <- ALLGO2ENS %>% 
  group_by(go_id) %>% 
  tally %>% 
  filter(n == 1) %>%
  extract2("go_id")
```

- This object contained all `r comma(nrow(ALLGO2ENS))` possible gene to GO mappings for this build of Ensembl.
- After building the database once, this object then became the reference object for all downstream analysis.
- A subset of `r length(singleGeneGO)` GO terms which mapped to only a single gene were also noted for exclusion from testing, as no significant results would be possible from these terms.

### Defining GO Term Levels

GO Terms were then defined in terms of their minimum distance to the Root Nodes, down to those requiring up to 3 steps.

```{r setGoLevels, cache=TRUE}
firstLevelGO <- list(
  BP = c(rootGO$BP, get(rootGO$BP, GOBPCHILDREN)),
  CC = c(rootGO$CC, get(rootGO$CC, GOCCCHILDREN)),
  MF = c(rootGO$MF, get(rootGO$MF, GOMFCHILDREN))
) %>%
  lapply(unique)
secondLevelGO <- list(
  BP = c(firstLevelGO$BP, lapply(firstLevelGO$BP, get, GOBPCHILDREN)),
  CC = c(firstLevelGO$CC, lapply(firstLevelGO$CC, get, GOCCCHILDREN)),
  MF = c(firstLevelGO$MF, lapply(firstLevelGO$MF, get, GOMFCHILDREN))
) %>%
  lapply(unlist) %>%
  lapply(unique) %>%
  lapply(function(x){x[!is.na(x)]})
thirdLevelGO <- list(
  BP = c(secondLevelGO$BP, lapply(secondLevelGO$BP, get, GOBPCHILDREN)),
  CC = c(secondLevelGO$CC, lapply(secondLevelGO$CC, get, GOCCCHILDREN)),
  MF = c(secondLevelGO$MF, lapply(secondLevelGO$MF, get, GOMFCHILDREN))
) %>%
  mclapply(unlist, mc.cores = nCores) %>%
  mclapply(unique, mc.cores = nCores) %>%
  mclapply(function(x){x[!is.na(x)]}, mc.cores = nCores)
```

## SNPs

```{r loadResults}
resultFiles <- file.path("..", "results", c("flkResults.tsv", 
                                            "alleleResults.tsv",
                                            "genotypeResults.tsv")) %>%
  set_names(c("flk", "alleles", "genotypes"))
results <- resultFiles %>%
  lapply(read_tsv) 
```

```{r resultsGR}
resultsGR <- full_join(
  results$flk %>%
    dplyr::select(snpID, Chr, BP, SNP, `Gum Creek (1996)`, `Oraparinna (2012)`,
                  FLK_p = F.LK.p.val, FLK_FDR = FDR),
  results$alleles %>%
    dplyr::select(snpID, Chr, BP, alleles_p = p, alleles_FDR = FDR)
) %>%
  full_join(
  results$genotypes %>%
    dplyr::select(snpID, Chr, BP, genotypes_p = p, genotypes_FDR = FDR)
  ) %>%
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE, 
    ignore.strand = TRUE,
    seqnames.field = "Chr", 
    start.field = "BP", 
    end.field = "BP", 
    seqinfo = seqinfo(ensGenes)) %>%
  sort()
```

```{r resultsGR_Sig}
resultsGR$Sig <- with(resultsGR, FLK_FDR < 0.05 | genotypes_FDR < 0.1)
```


# Relationship of SNPs to Genes

## SNPs within genes

```{r snpInGenes}
snpInGenes <- resultsGR %>% 
  subset(Sig) %>%
  findOverlaps(ensGenes)
snpInExons <- resultsGR %>% 
  subset(Sig) %>%
  findOverlaps(ensExons)
```

A total of `r length(snpInGenes)` significant SNPs were found to be within the start and end points of genes.
`r length(snpInExons)` of these were within the coding regions of these genes.

```{r, echo=FALSE}
ensGenes[subjectHits(snpInGenes)] %>%
  mcols() %>%
  as.data.frame() %>%
  dplyr::select(gene_id, Name) %>%
  bind_cols(
    resultsGR %>%
      subset(Sig) %>%
      magrittr::extract(queryHits(snpInGenes)) %>%
      as.data.frame() %>%
      dplyr::select(Chr = seqnames, BP = start, snpID)
  ) %>%
  mutate(inExon = snpID %in% subset(resultsGR, Sig)[queryHits(snpInExons)]$snpID) %>%
  rowwise() %>%
  mutate(snpID = if_else(inExon, paste0(snpID,"*"), snpID),
         Name = ifelse(is.na(Name), "", Name)) %>%
  dplyr::select(`Gene ID` = gene_id, Name, Chr, BP, snpID) %>%
  pander(justify = "llrrl",
         style = "rmarkdown",
         big.mark = ",",
         split.table = Inf,
         caption = paste("Genes with significant SNPs located between their start and end positions.", "SNPs within exons are indicated with an additional asterisk."))
```

- It was also noted that proteins produced from `MYO1B` and `USP46`are known interacting proteins in human biology

## Genes within 20kb

```{r allGenesIn20kb}
allGenesIn20kb <- subsetByOverlaps(
  ensGenes, 
  resultsGR %>%
    resize(width = 40001,fix = "center") %>%
    trim()
    )
sigGenesIn20kb <- subsetByOverlaps(
  ensGenes, 
  resultsGR %>%
    subset(Sig)%>%
    resize(width = 40001,fix = "center") %>%
    trim()
    )
```

As a an initial conservative approach, the set of genes within 20kb of a SNP was investigated.
This produced a list of `r length(allGenesIn20kb)` genes in total, of which `r length(sigGenesIn20kb)` overlapped SNPs of interest.


```{r go2Test20kb}
go2Test20kb <- ALLGO2ENS %>%
  filter(ensembl_gene_id %in% sigGenesIn20kb$gene_id,
         !go_id %in% unlist(thirdLevelGO)) %>%
  extract2("go_id") %>%
  unique %>%
  setdiff(singleGeneGO)
```

A set of `r length(go2Test20kb)` GO terms mapping to more than one gene, and with at least 4 steps back to the ontology roots, were assigned to the genes within 20kb of the `r sum(resultsGR$Sig)` candidate SNPs.
These were then tested for enrichment using the set of genes within 20kb of the remaining `r sum(!resultsGR$Sig)` SNPs

Due to the closely related nature of some SNPs in this dataset the possibility of two SNPs being within 20kb of the same genes was extremely high, and this approach would ensure that each gene only appears once despite the possibility of mapping to multiple SNPs within the dataset.

Fisher's Exact Test was then used to test for enrichment by counting the genes mapped to each GO term within the set of significant, and non-significant SNPs, and comparing to the genes not mapped to each GO term in both sets of SNPs.
All p values were then adjusted using Benjamini-Hochberg's FDR.


```{r goResults20kb}
nSigGenes <- length(sigGenesIn20kb)
nNotSigGenes <- length(allGenesIn20kb) - nSigGenes
goResults20kb <- go2Test20kb %>%
  lapply(function(x){
    # Form a matrix of the counts
    mat <- filter(ALLGO2ENS, go_id == x) %>%
      mutate(Sig = ensembl_gene_id %in% sigGenesIn20kb$gene_id) %>%
      acast(Sig~., fun.aggregate = length, value.var = "Sig")
    # Ensure it is a matrix, then set row/colnames
    if (length(mat) == 1) mat <- c(0, mat)
    mat <- cbind(mat, withoutGO = c(nNotSigGenes, nSigGenes) - mat)
    colnames(mat)[1] <- "withGO"
    rownames(mat) <- c("notNearSNP", "nearSNP")
    # Fisher's exact test
    ft <- fisher.test(mat)
    data_frame(go_id = x,
               expected = nSigGenes * mat["notNearSNP", "withGO"] / nNotSigGenes,
               observed = mat["nearSNP", "withGO"],
               p = ft$p.value)
  }) %>%
  bind_rows() %>%
  mutate(FDR = p.adjust(p, "fdr")) %>%
  arrange(p)
```

A total of `r nrow(filter(goResults20kb, observed > expected, FDR < 0.05))` GO terms were considered as enriched using the criteria of an FDR-adjusted p-value < 0.05 and with observed numbers greater than that predicted by the ratio in the non-significant SNP genes.

```{r, echo=FALSE}
goResults20kb %>%
  filter(observed > expected, FDR < 0.05) %>%
  mutate(Term = Term(go_id),
         Ont = Ontology(go_id)) %>%
  set_names(str_to_title(names(.))) %>%
  dplyr::select(`GO ID` = Go_id, Term, Ont, Observed, Expected, p = P, FDR = Fdr) %>%
  pander(justify = "lllrrrr",
         style = "rmarkdown",
         split.table = Inf,
         caption = paste("Gene Ontologies considered as enriched amongst the set of",
                         length(sigGenesIn20kb),
                         "genes within 20kb of the significant SNPs.",
                         "The number of genes matching each term is given in the 'Observed' column."))
```

## Genes within 40kb

```{r allGenesIn40kb}
allGenesIn40kb <- subsetByOverlaps(
  ensGenes, 
  resultsGR %>%
    resize(width = 80001,fix = "center") %>%
    trim()
    )
sigGenesIn40kb <- subsetByOverlaps(
  ensGenes, 
  resultsGR %>%
    subset(Sig)%>%
    resize(width = 80001,fix = "center") %>%
    trim()
    )
```

The same approach was then repeated for genes within 40kb of each SNP.
This produced a list of `r length(allGenesIn40kb)` genes in total, of which `r length(sigGenesIn40kb)` overlapped SNPs of interest.


```{r go2Test40kb}
go2Test40kb <- ALLGO2ENS %>%
  filter(ensembl_gene_id %in% sigGenesIn40kb$gene_id,
         !go_id %in% unlist(thirdLevelGO)) %>%
  extract2("go_id") %>%
  unique %>%
  setdiff(singleGeneGO)
```

A set of `r length(go2Test40kb)` GO terms mapping to more than one gene, and with at least 4 steps back to the ontology roots were assigned to the genes within 40kb of the `r sum(resultsGR$Sig)` candidate SNPs.
These were then tested for enrichment using the set of genes within 40kb of the remaining `r sum(!resultsGR$Sig)` SNPs


```{r goResults40kb}
nSigGenes <- length(sigGenesIn40kb)
nNotSigGenes <- length(allGenesIn40kb) - nSigGenes
goResults40kb <- go2Test40kb %>%
  lapply(function(x){
    # Form a matrix of the counts
    mat <- filter(ALLGO2ENS, go_id == x) %>%
      mutate(Sig = ensembl_gene_id %in% sigGenesIn40kb$gene_id) %>%
      acast(Sig~., fun.aggregate = length, value.var = "Sig")
    # Ensure it is a matrix, then set row/colnames
    if (length(mat) == 1) mat <- c(0, mat)
    mat <- cbind(mat, withoutGO = c(nNotSigGenes, nSigGenes) - mat)
    colnames(mat)[1] <- "withGO"
    rownames(mat) <- c("notNearSNP", "nearSNP")
    # Fisher's exact test
    ft <- fisher.test(mat)
    data_frame(go_id = x,
               expected = nSigGenes * mat["notNearSNP", "withGO"] / nNotSigGenes,
               observed = mat["nearSNP", "withGO"],
               p = ft$p.value)
  }) %>%
  bind_rows() %>%
  mutate(FDR = p.adjust(p, "fdr")) %>%
  arrange(p)
```

A total of `r nrow(filter(goResults40kb, observed > expected, FDR < 0.05))` GO terms were considered as enriched using the criteria of an FDR-adjusted p-value < 0.05 and with observed numbers greater than that predicted by the ratio in the non-significant SNP genes.

```{r, echo=FALSE}
goResults40kb %>%
  filter(observed > expected, FDR < 0.05) %>%
  mutate(Term = Term(go_id),
         Ont= Ontology(go_id)) %>%
  set_names(str_to_title(names(.))) %>%
  dplyr::select(`GO ID` = Go_id, Term, Ont, Observed, Expected, p = P, FDR = Fdr) %>%
  pander(justify = "lllrrrr",
         style = "rmarkdown",
         split.table = Inf,
         caption = paste("Gene Ontologies considered as enriched amongst the set of",
                         length(sigGenesIn40kb),
                         "genes within 40kb of the significant SNPs.",
                         "The number of genes matching each term is given in the 'Observed' column."))
```

Similar sets of terms were detected at both 20kb and 40kb.
The appearance of terms connected to the Zinc ion may be of note as the link between zinc and clearance of HCV has recently been established, via the IFN-&#947; pathway

The genes associated with each of these GO terms is given below:

```{r, echo=FALSE}
goResults40kb %>%
    filter(observed > expected, FDR < 0.05) %>% 
  dplyr::select(go_id) %>% 
  left_join(ALLGO2ENS) %>% 
  mutate(Term = Term(go_id)) %>% 
  left_join(as.data.frame(ensGenes),
            by = c("ensembl_gene_id" = "gene_id")) %>%
  dplyr::select(go_id, Term, ensembl_gene_id, Name) %>%
  as.data.frame() %>% 
  mutate(go_id = ifelse(duplicated(go_id), "", go_id),
         Term = ifelse(duplicated(Term), "", Term),
         Name = ifelse(is.na(Name), "", Name)) %>%
  dplyr::rename(`GO ID` = go_id,
         `Gene ID` = ensembl_gene_id) %>%
  pander(justify = "llll",
         split.cells = Inf,
         split.table = Inf,
         caption = "Genes associated wth each enriched GO term")
```


### Export of Results

The set of genes within 40kb of each SNP of interest were then exported.

```{r hitsIn40kb}
hitsIn40kb <- findOverlaps(
  resultsGR %>%
    subset(Sig)%>%
    resize(width = 80001,fix = "center") %>%
    trim(),
    ensGenes
    )
```

```{r exportGenesIn40kb}
tsvOut <- file.path("..", "results", "GenesIn40kb.tsv")
list(
  resultsGR %>%
    subset(Sig) %>%
    extract(queryHits(hitsIn40kb)) %>%
    as.data.frame() %>%
    dplyr::select(snpID, Chr = seqnames, BP = start, genotypes_p,  FLK_p) ,
  ensGenes %>%
    extract(subjectHits(hitsIn40kb)) %>%
    as.data.frame() %>%
    dplyr::select(GeneStart = start, GeneEnd = end, strand, NearGene = gene_id, GeneName = Name) 
) %>%
  bind_cols %>%
  as_data_frame() %>%
    rowwise()%>%
  mutate(LocusID = gsub("([0-9]*)_[0-9]*", "\\1", snpID),
         dist2Gene = if_else(BP > GeneStart && BP < GeneEnd, 0L, min(abs(c(BP - GeneStart, BP - GeneEnd)))),
         GeneWidth = abs(GeneStart - GeneEnd)) %>%
  dplyr::select(LocusID, snpID, Chr, BP, NearGene, GeneName, GeneStrand = strand, GeneStart, GeneEnd, GeneWidth, dist2Gene, genotypes_p, FLK_p) %>%
  mutate(minP = min(genotypes_p, FLK_p)) %>%
  arrange(minP) %>%
  dplyr::select(-minP) %>%
  as.data.frame() %>%
  write_tsv(tsvOut)
```

```{r sessionInfo, results='asis'}
pander(sessionInfo()) 
```
